src/Test/Utilities/Desktop/InstrumentationChecker.cs

```diff
@@ -0,0 +1,398 @@
+            {
+                var actualSnippets = GetActualSnippets(method, reader, sourceLines);
+                var expectedSnippets = _spanExpectations[method].SnippetExpectations;
+
+                AssertEx.Equal(expectedSnippets, actualSnippets, new SnippetComparer(), $"Validation of method {method} failed.");
+            }
+        }
+
+        // Fails the test and outputs a template for updating the test code with appropriate expectations.
+        internal abstract void Dump(DynamicAnalysisDataReader reader, string[] sourceLines);
+
+        internal abstract void AddConsoleExpectation(string text);
+
+        internal static ImmutableArray<string> GetActualSnippets(int method, DynamicAnalysisDataReader reader, string[] sourceLines)
+        {
+            var actualSpans = reader.GetSpans(reader.Methods[method - 1].Blob);
+
+            return actualSpans.SelectAsArray((span, lines) =>
+            {
+                if (span.StartLine >= lines.Length) { return null; }
+                return lines[span.StartLine].Substring(span.StartColumn).TrimEnd(new[] { '\r', '\n', ' ' });
+            }, sourceLines);
+        }
+
+        public class MethodChecker
+        {
+            private List<string> _snippetExpectations;
+            private BaseInstrumentationChecker _checker;
+
+            public MethodChecker(BaseInstrumentationChecker checker, bool noSnippets = false)
+            {
+                _checker = checker;
+
+                if (!noSnippets)
+                {
+                    _snippetExpectations = new List<string>();
+                }
+            }
+
+            public string[] SnippetExpectations { get { return _snippetExpectations.ToArray(); } }
+
+            /// <summary>
+            /// Records the expectation that the following span will be covered and resembles the provided snippet.
+            /// </summary>
+            public MethodChecker True(string expectedSourceSnippet = null)
+            {
+                return Expect("True", expectedSourceSnippet);
+            }
+
+            /// <summary>
+            /// Records the expectation that the following span will *not* be covered and resembles the provided snippet.
+            /// </summary>
+            public MethodChecker False(string expectedSourceSnippet = null)
+            {
+                return Expect("False", expectedSourceSnippet);
+            }
+
+            private MethodChecker Expect(string text, string expectedSourceSnippet)
+            {
+                _checker.AddConsoleExpectation(text);
+                if (_snippetExpectations != null)
+                {
+                    _snippetExpectations.Add(expectedSourceSnippet);
+                }
+                else
+                {
+                    Assert.True(expectedSourceSnippet == null,
+                        "You must pass a snippet when checking the method with M if you intend to verify snippets within the method.");
```

Comment:
Is the assert text correct? If the assert fails, a (non-null) snippet was passed in. Also, should `M` be `Method`?